src/Compilers/Core/Portable/Compilation/Expression.cs

```diff
@@ -0,0 +1,365 @@
+                    return ConstantValue.Create(value);
+                case SpecialType.System_Int64:
+                    return ConstantValue.Create((long)value);
+                case SpecialType.System_UInt32:
+                    return ConstantValue.Create((uint)value);
+                case SpecialType.System_UInt64:
+                    return ConstantValue.Create((ulong)value);
+                case SpecialType.System_UInt16:
+                    return ConstantValue.Create((ushort)value);
+                case SpecialType.System_Int16:
+                    return ConstantValue.Create((short)value);
+                case SpecialType.System_SByte:
+                    return ConstantValue.Create((sbyte)value);
+                case SpecialType.System_Byte:
+                    return ConstantValue.Create((byte)value);
+                case SpecialType.System_Char:
+                    return ConstantValue.Create((char)value);
+                case SpecialType.System_Boolean:
+                    return ConstantValue.Create(value != 0 ? true : false);
+                case SpecialType.System_Single:
+                    return ConstantValue.Create((float)value);
+                case SpecialType.System_Double:
+                    return ConstantValue.Create((double)value);
+                case SpecialType.System_Object:
+                    return ConstantValue.Create(1, ConstantValueTypeDiscriminator.Int32);
+            }
+
+            if (type.TypeKind == TypeKind.Enum)
+            {
+                return SynthesizeNumeric(((INamedTypeSymbol)type).EnumUnderlyingType, value);
+            }
+
+            return null;
+        }
+
+        public static BinaryOperationKind DeriveAdditionKind(ITypeSymbol type)
+        {
+            switch (type.SpecialType)
+            {
+                case SpecialType.System_Int32:
+                case SpecialType.System_Int64:
+                case SpecialType.System_Int16:
+                case SpecialType.System_SByte:
+                    return BinaryOperationKind.IntegerAdd;
+                case SpecialType.System_UInt32:
+                case SpecialType.System_UInt64:
+                case SpecialType.System_UInt16:
+                case SpecialType.System_Byte:
+                case SpecialType.System_Char:
+                case SpecialType.System_Boolean:
+                    return BinaryOperationKind.UnsignedAdd;
+                case SpecialType.System_Single:
+                case SpecialType.System_Double:
+                    return BinaryOperationKind.FloatingAdd;
+                case SpecialType.System_Object:
+                    return BinaryOperationKind.ObjectAdd;
+            }
+
+            if (type.TypeKind == TypeKind.Enum)
+            {
+                return Semantics.BinaryOperationKind.EnumAdd;
+            }
+
+            return Semantics.BinaryOperationKind.None;
+        }
+
+        public static LiteralKind DeriveLiteralKind(ITypeSymbol type)
```

Comment:
LIteralKind is being removed, because SpecialType is available from a literal's ResultType.